prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -27,6 +30,157 @@ def check_for_location_conflicts(substations):
+    to_sub_missing = line2sub["TO_SUB"].isnull().values
+
+    unique_endpoint_coord = set(
+        line2sub["FROM"].unique().tolist() + line2sub["TO"].unique().tolist()
+    )
+
+    n_two_sub_missing = sum(from_sub_missing & to_sub_missing)
+    n_one_sub_missing = sum(from_sub_missing ^ to_sub_missing)
+    n_unique_orphan_endpoint = len(unique_endpoint_coord - set(subcoord2subid))
+    n_unconnected_sub = len(set(subcoord2subid) - unique_endpoint_coord)
+
+    print(
+        f"there are {n_two_sub_missing} lines with two substations missing\n"
+        f"there are {n_one_sub_missing} lines with one substation missing\n"
+        f"there are {n_unique_orphan_endpoint} unique orphan endpoint\n"
+        f"there are {n_unconnected_sub} substations unconnected\n"
+    )
+
+    return line2sub
+
+
+def assign_substations_to_lines(substations, lines, line2sub):
+    """Locate closest substation of unconnected lines and update information
+    information of lines. Note that the ``line2sub`` data frame is modified in-place.
+
+    :param pandas.DataFrame substations: data frame of substations.
+    :param pandas.DataFrame lines: data frame of all lines.
+    :param pandas.DataFrame line2sub: indices: table mapping lines to substations as
+        returned by :func:`map_lines_to_substations_using_coords`.
+    :return: (*pandas.DataFrame*) -- original lines data frame with updated *SUB_1*
+        and *SUB_2* columns, new *SUB_1_ID*, *SUB_2_ID* along with *OTHER_SUB_1_ID*
+        and *OTHER_SUB_2_ID* when multiple location are associated to the same endpoint.
+    """
+    print("finding closest neighbor to unconnected lines")
+
+    # find closest neighbor of unconnected lines
+    subcoord2subid = substations.groupby(["LONGITUDE", "LATITUDE"]).groups
+    subcoord = list(subcoord2subid)
+    end_sub = {"FROM": "FROM_SUB", "TO": "TO_SUB"}
+    missing_points = set().union(
+        *[set(line2sub[e].loc[line2sub[e_sub].isna()]) for e, e_sub in end_sub.items()]
+    )
+    endpoint2neighbor = {
+        point: subcoord2subid[subcoord[find_closest_neighbor(point, subcoord)]]
+        for point in tqdm(missing_points, total=len(missing_points))
+    }
+    filled_subs = [
+        line2sub[e].loc[line2sub[e_sub].isna()].map(endpoint2neighbor).map(list)
+        for e, e_sub in end_sub.items()
+    ]
+    line2sub.update(pd.concat(filled_subs, axis=1).rename(columns=end_sub))
+
+    # select all substations
+    new_substations = substations.copy()
+    new_substations = substations.loc[
+        set(
+            itertools.chain(
+                *{tuple(i) for j in ["FROM_SUB", "TO_SUB"] for i in line2sub[j]}
+            )
+        )
+    ]
+
+    # pick unique substation at each end. Order of preference:
+    # SUBSTATION --> TAP --> RISER --> DEAD END
+    all2one = {
+        tuple(i): None
+        for j in ["FROM_SUB", "TO_SUB"]
+        for i in line2sub[j]
+        if len(i) > 1
+    }
```

Comment:
Is there a reason that this is a dict instead of a set or other iterable?

Reply:
No specific reason. I decided to use a dictionary to relate the multiple substations at the same location to a unique one. If you think it would be cleaner to have two iterables instead, I can refactor the code.

Reply:
It doesn't really matter; starting as a dict with dummy values and then filling them in vs. starting as a set and then iterating over that to define a dict should be basically equivalent. I just wanted to make sure I wasn't missing something.